### ISSUE TITLE
Function to find positions of a sub-matrix inside a matrix

### ISSUE THREAD

**Issue by 2005m** [Rdatatable/data.table#4114]
I do not know any package who has implemented a function that can find the positions (row and column index) of a small matrix inside a larger matrix. After discussing it with @mattdowle and @MichaelChirico a couple of month ago, they requested that I raise the question here. If it is ok with @Rdatatable/project-members , I would like to add such function/feature in `data.table`. Of course I will take care of the implementation.

**Comment by jangorecki** [comment]
I recall very smilar question on R-devel asking for that functionality in base R, few packages were provided in response there. If that is meant to be a function for `matrix` class I am not sure if data.table is best place for that.

**Comment by 2005m** [comment]
I actually did ask the question on r-dev and they were clearly not open to include such a general function to base R. Plus I contacted the maintainer of the `Matrix` package and he did not seem interested either. @jangorecki, note that such function would not only work on matrix but vector as well. Vectors are nothing else that else than matrix with one of their dimension equal to 1. I also would like to add that all functions in`data.table` are very general and could easily find a place in other packages. Take the example of `fifelse`, `fcase`, `fread`, `fwrite`, these were created because base R is not good enough. Also take the example of the sorting function in base R that was ported from `data.table`. The reality is that it is not easy to change core R, because there are a lot of "politics"...nor is it easy to get someone to include a function in a package even when you are offering to implement it.

**Comment by mattdowle** [comment]
Great points all. On fread and fwrite they were created in data.table to create a data.table by default, and also, eventually they can do some data.table-specific things inside them; e.g. they can create and store attributes on the data.table; in other words fread and fwrite are weakly integrated currently but there is some further tighter integration in mind to be done. fifelse and fcase are for use inside data.table queries and there's potential for optimization, so that's why those. Fast sort is needed for setkey() upon which fast join depends; so there are dependencies there.  Other ones are dcast and melt: they could have been enhancements to reshape2 or outside data.table, but one reason was their internals used existing data.table code at C level; so they are entwined with data.table at a code level. Having these code dependencies within the package makes the package self contained so there is no chance of different versions having to work together which have never been tested before; the test suite stamps the single version of the package as a whole as passing tests.  And all the functions mentioned so far have been about data.frame|table.

That said, I'd very much welcome this feature (finding sub-matrix part of matrix) to be part of data.table. Just because it sounds cool, you want to, and I guess it would likely be reusing data.table internals at C level like forder and bmerge, perhaps. There is matrix-join like logic in foverlaps which could be reused or built upon or something, perhaps.  I don't know how advanced your thoughts are on implementation. I saw that fcase was advanced good C code, so we want to encourage you to be part of the project and do whatever you'd like within it!

Could you write a few words about which fields need this functionality and why? It helps motivation levels to know that.

**Comment by 2005m** [comment]
Thanks @mattdowle. Yes I will write it in C. I will provide the code in the new year. I have a version which is written in R only and is surprisingly fast but the C version will be faster. The first time I have used this function was in a project where I needed to create some kind of "robot" using the java Robot class which enable to control mouse and keyboard entries. All this was called from R. The "robot" was taking screenshot of someone's desktop. The screenshot is then saved as an image. (An image is nothing else than a matrix) I was then using this matrix find function to identify where specific applications/elements were on the screen; I was then able to manipulate them with Java robot.

This function can be somehow (but not really) seen as a generalisation of the `which` function that identifies the position of a single element in a vector or matrix. However here I push it a step futher, I can identify the position of a vector inside another vector or matrix inside another one. I can choose to take only the first occurrence or return all of them. I can filter out the ones that are overlapping.

Here is an example of functions that find the position of vector inside a larger one:

https://coolbutuseless.bitbucket.io/2018/04/03/finding-a-length-n-needle-in-a-haystack/

I am planning to push the functions written in the above post a step further (i.e. make the functionnality available at matrix level) and would like go even faster than the fastest function.
